wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1064 @@
+        iotsafe_cmd_add_tlv(csim_cmd, IOTSAFE_TAG_PUBKEY_ID, 1, &pubkey_idx);
+        iotsafe_cmd_complete(csim_cmd);
+        ret = expect_csim_response(csim_cmd, (word32)strlen(csim_cmd), &resp);
+        if (ret < 1) {
+            WOLFSSL_MSG("Unexpected reply in DH command");
+            return -1;
+        }
+        *outlen = hex_to_bytes(resp, out, ret / 2);
+        ret = 0;
+    } else {
+        ecc_key*  privKey = NULL;
+        ecc_key*  pubKey = NULL;
+        ret = wc_ecc_init(&tmpKey);
+        if (ret != 0) {
+            return ret;
+        }
+        /* for client: create and export public key */
+        if (side == WOLFSSL_CLIENT_END) {
+            WC_RNG rng;
+            privKey = &tmpKey;
+            pubKey = otherKey;
+            ret = wc_InitRng(&rng);
+            if (ret == 0) {
+                ret = wc_ecc_make_key_ex(&rng, 0, privKey, otherKey->dp->id);
+                if (ret == 0)
+                    ret = wc_ecc_export_x963(privKey, pubKeyDer, pubKeySz);
+                wc_FreeRng(&rng);
+            }
+        }
+        /* for server: import public key */
+        else if (side == WOLFSSL_SERVER_END) {
+            privKey = otherKey;
+            pubKey = &tmpKey;
+            ret = wc_ecc_import_x963_ex(pubKeyDer, *pubKeySz, pubKey,
+                    otherKey->dp->id);
+        }
+        else {
+            ret = BAD_FUNC_ARG;
+        }
+        /* generate shared secret and return it */
+        if (ret == 0) {
+            ret = wc_ecc_shared_secret(privKey, pubKey, out, outlen);
+        }
+        wc_ecc_free(&tmpKey);
+    }
+    return ret;
+}
+
+#ifndef NO_DH
+/* Public DH Key Callback support */
+static int wolfIoT_dh_agree(WOLFSSL* ssl, struct DhKey* key,
+        const unsigned char* priv, unsigned int privSz,
+        const unsigned char* otherPubKeyDer, unsigned int otherPubKeySz,
+        unsigned char* out, word32* outlen,
+        void* ctx)
+{
+
+    WOLFSSL_MSG("IOTSAFE: Called wolfIoT_ecc_dh_agree\n");
+    (void)ctx;
+    (void)ssl;
+    (void)key;
+    (void)priv;
+    (void)privSz;
+    (void)otherPubKeyDer;
+    (void)otherPubKeySz;
+    (void)out;
+    (void)outlen;
+
+    /* TODO */
+    return 0;
+}
+#endif /* ! NO_DH */
+
+#endif /* HAVE_PK_CALLBACKS */
+
+/* Public API for ecc */
+
+int ecc_import_public_iotsafe(ecc_key *key, byte key_id)
+{
+    return iotsafe_get_public_key(key_id, key);
+}
+
+int ecc_export_public_iotsafe(ecc_key *key, byte key_id)
+{
+    return iotsafe_put_public_key(key_id, key);
+}
+
+int ecc_sign_hash_iotsafe(byte *in, word32 inlen, byte *out,
+        word32 *outlen, byte key_id)
+{
+
+    return iotsafe_sign_hash(key_id, IOTSAFE_HASH_SHA256, IOTSAFE_SIGN_ECDSA,
+            in, inlen, out, outlen);
+}
+int ecc_verify_hash_iotsafe(byte *sig, word32 siglen, byte *hash,
+        word32 hashlen, int *res, byte key_id)
+{
+    word32 r_size = IOTSAFE_ECC_KSIZE;
+    word32 s_size = IOTSAFE_ECC_KSIZE;
+    *res = 0;
```

Comment:
All variables should be declared at top of function. Move `*res=0` down a line.

Reply:
spotted one more of these as well.